<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>filter annotation for plotting coverages — filter_annotation • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="filter annotation for plotting coverages — filter_annotation"><meta name="description" content="Removes isoform annotations that could produce ambigious reads, such as isoforms
that only differ by the 5' / 3' end. This could be useful for plotting average
coverage plots."><meta property="og:description" content="Removes isoform annotations that could produce ambigious reads, such as isoforms
that only differ by the 5' / 3' end. This could be useful for plotting average
coverage plots."><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>filter annotation for plotting coverages</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/model_decay.R" class="external-link"><code>R/model_decay.R</code></a></small>
      <div class="d-none name"><code>filter_annotation.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Removes isoform annotations that could produce ambigious reads, such as isoforms
that only differ by the 5' / 3' end. This could be useful for plotting average
coverage plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_annotation</span><span class="op">(</span><span class="va">annotation</span>, keep <span class="op">=</span> <span class="st">"tss_differ"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-annotation">annotation<a class="anchor" aria-label="anchor" href="#arg-annotation"></a></dt>
<dd><p>path to the GTF annotation file, or the parsed GenomicRanges
object.</p></dd>


<dt id="arg-keep">keep<a class="anchor" aria-label="anchor" href="#arg-keep"></a></dt>
<dd><p>string, one of 'tss_differ' (only keep isoforms that all differ
by the transcription start site position), 'tes_differ' (only keep those that
differ by the transcription end site position), 'both' (only keep those that
differ by both the start and end site), or 'single_transcripts' (only keep
genes that contains a sinlge transcript).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>GenomicRanges of the filtered isoforms</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">filtered_annotation</span> <span class="op">&lt;-</span> <span class="fu">filter_annotation</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata/rps24.gtf.gz'</span>, package <span class="op">=</span> <span class="st">'FLAMES'</span><span class="op">)</span>, keep <span class="op">=</span> <span class="st">'tes_differ'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The "phase" metadata column contains non-NA values for features of type</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   stop_codon. This information was ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"><span><span class="va">filtered_annotation</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 6 ranges and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand |     tx_id              tx_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;          &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]    chr14   19-5159      + |         1 ENSMUST00000225994.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]    chr14   32-3389      + |         6 ENSMUST00000225117.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]    chr14   68-5124      + |         7 ENSMUST00000224568.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]    chr14   86-1118      + |         8 ENSMUST00000224549.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]    chr14  160-2761      + |         9 ENSMUST00000224569.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]    chr14  450-1290      + |        10 ENSMUST00000224699.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

