<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>bulk variant identification — find_variants • FLAMES</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bulk variant identification — find_variants"><meta name="description" content="Treat each bam file as a bulk sample and identify variants against the reference"><meta property="og:description" content="Treat each bam file as a bulk sample and identify variants against the reference"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>bulk variant identification</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/sc_mutations.R" class="external-link"><code>R/sc_mutations.R</code></a></small>
      <div class="d-none name"><code>find_variants.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Treat each bam file as a bulk sample and identify variants against the reference</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">find_variants</span><span class="op">(</span></span>
<span>  <span class="va">bam_path</span>,</span>
<span>  <span class="va">reference</span>,</span>
<span>  <span class="va">annotation</span>,</span>
<span>  min_nucleotide_depth <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  homopolymer_window <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  annotated_region_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  names_from <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam-path">bam_path<a class="anchor" aria-label="anchor" href="#arg-bam-path"></a></dt>
<dd><p>character(1) or character(n): path to the bam file(s) aligned to the
reference genome (NOT the transcriptome!).</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>DNAStringSet: the reference genome</p></dd>


<dt id="arg-annotation">annotation<a class="anchor" aria-label="anchor" href="#arg-annotation"></a></dt>
<dd><p>GRanges: the annotation of the reference genome. You can load
a GTF/GFF annotation file with <code>anno &lt;- rtracklayer::import(file)</code>.</p></dd>


<dt id="arg-min-nucleotide-depth">min_nucleotide_depth<a class="anchor" aria-label="anchor" href="#arg-min-nucleotide-depth"></a></dt>
<dd><p>integer(1): minimum read depth for a position to be
considered a variant.</p></dd>


<dt id="arg-homopolymer-window">homopolymer_window<a class="anchor" aria-label="anchor" href="#arg-homopolymer-window"></a></dt>
<dd><p>integer(1): the window size to calculate the homopolymer
percentage. The homopolymer percentage is calculated as the percentage of the most
frequent nucleotide in a window of <code>-homopolymer_window</code> to <code>homopolymer_window</code>
nucleotides around the variant position, excluding the variant position itself.
Calculation of the homopolymer percentage is skipped when <code>homopolymer_window = 0</code>.
This is useful for filtering out Nanopore sequencing errors in homopolymer regions.</p></dd>


<dt id="arg-annotated-region-only">annotated_region_only<a class="anchor" aria-label="anchor" href="#arg-annotated-region-only"></a></dt>
<dd><p>logical(1): whether to only consider variants outside
annotated regions. If <code>TRUE</code>, only variants outside annotated regions will be
returned. If <code>FALSE</code>, all variants will be returned, which could take significantly
longer time.</p></dd>


<dt id="arg-names-from">names_from<a class="anchor" aria-label="anchor" href="#arg-names-from"></a></dt>
<dd><p>character(1): the column name in the metadata column of the annotation
(<code>mcols(annotation)[, names_from]</code>) to use for the <code>region</code> column in the output.</p></dd>


<dt id="arg-threads">threads<a class="anchor" aria-label="anchor" href="#arg-threads"></a></dt>
<dd><p>integer(1): number of threads to use. Threading is done over each
annotated region and (if <code>annotated_region_only = FALSE</code>) unannotated gaps for
each bam file.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A tibble with columns: seqnames, pos, nucleotide, count, sum, freq, ref, region,
homopolymer_pct, bam_path The homopolymer percentage is calculated as the percentage of the
most frequent nucleotide in a window of <code>homopolymer_window</code> nucleotides around
the variant position, excluding the variant position itself.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Each bam file is treated as a bulk sample to perform pileup and identify variants.
You can run <code>sc_mutations</code> with the variants identified with this function
to get single-cell allele counts. Note that reference genome FASTA files may have
the chromosome names field as `&gt;chr1 1` instead of `&gt;chr1`. You may need to remove
the trailing number to match the chromosome names in the bam file, for example with
<code>names(ref) &lt;- sapply(names(ref), function(x) strsplit(x, " ")[[1]][1])</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"rps24.fa"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.fa.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>, destname <span class="op">=</span> <span class="va">genome_fa</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/mritchielab/FLAMES/devel/tests/testthat/demultiplexed.fq"</span>,</span></span>
<span class="r-in"><span>  destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultipelxed.fq"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="co"># can't be bothered to run demultiplexing again</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="sys_which.html">sys_which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="minimap2_align.html">minimap2_align</a></span><span class="op">(</span> <span class="co"># align to genome</span></span></span>
<span class="r-in"><span>    config <span class="op">=</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/SIRV_config_default.json"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    fa_file <span class="op">=</span> <span class="va">genome_fa</span>,</span></span>
<span class="r-in"><span>    fq_in <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"demultipelxed.fq"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    annot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.gtf.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    outdir <span class="op">=</span> <span class="va">outdir</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu">find_variants</span><span class="op">(</span></span></span>
<span class="r-in"><span>    bam_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"align2genome.bam"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    reference <span class="op">=</span> <span class="va">genome_fa</span>,</span></span>
<span class="r-in"><span>    annotation <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr14"</span>, <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    min_nucleotide_depth <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

