<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get read coverages from BAM file — get_coverage • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get read coverages from BAM file — get_coverage"><meta name="description" content="Get the read coverages for each transcript in the BAM file (or a GAlignments object).
The read coverages are sampled at 100 positions along the transcript, and the coverage is scaled
by dividing the coverage at each position by the total read counts for the transcript. If a BAM
file is provided, alignment with MAPQ &amp;lt; 5, secondary alignments and supplementary alignments
are filtered out. A GAlignments object can also be provided in case alternative filtering
is desired."><meta property="og:description" content="Get the read coverages for each transcript in the BAM file (or a GAlignments object).
The read coverages are sampled at 100 positions along the transcript, and the coverage is scaled
by dividing the coverage at each position by the total read counts for the transcript. If a BAM
file is provided, alignment with MAPQ &amp;lt; 5, secondary alignments and supplementary alignments
are filtered out. A GAlignments object can also be provided in case alternative filtering
is desired."><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Get read coverages from BAM file</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/model_decay.R" class="external-link"><code>R/model_decay.R</code></a></small>
      <div class="d-none name"><code>get_coverage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Get the read coverages for each transcript in the BAM file (or a GAlignments object).
The read coverages are sampled at 100 positions along the transcript, and the coverage is scaled
by dividing the coverage at each position by the total read counts for the transcript. If a BAM
file is provided, alignment with MAPQ &lt; 5, secondary alignments and supplementary alignments
are filtered out. A <code>GAlignments</code> object can also be provided in case alternative filtering
is desired.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_coverage</span><span class="op">(</span><span class="va">bam</span>, min_counts <span class="op">=</span> <span class="fl">10</span>, remove_UTR <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">annotation</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>path to the BAM file, or a parsed GAlignments object</p></dd>


<dt id="arg-min-counts">min_counts<a class="anchor" aria-label="anchor" href="#arg-min-counts"></a></dt>
<dd><p>numeric, the minimum number of alignments required for a transcript to be included</p></dd>


<dt id="arg-remove-utr">remove_UTR<a class="anchor" aria-label="anchor" href="#arg-remove-utr"></a></dt>
<dd><p>logical, if <code>TRUE</code>, remove the UTRs from the coverage</p></dd>


<dt id="arg-annotation">annotation<a class="anchor" aria-label="anchor" href="#arg-annotation"></a></dt>
<dd><p>(Required if <code>remove_UTR = TRUE</code>) path to the GTF annotation file</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a tibble of the transcript information and coverages, with the following columns:</p><ul><li><p><code>transcript</code>: the transcript name / ID</p></li>
<li><p><code>read_counts</code>: the total number of aligments for the transcript</p></li>
<li><p><code>coverage_1-100</code>: the coverage at each of the 100 positions along the transcript</p></li>
<li><p><code>tr_length</code>: the length of the transcript</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create a BAM file with minimap2_realign</span></span></span>
<span class="r-in"><span><span class="va">temp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">temp_path</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">file_url</span> <span class="op">&lt;-</span> <span class="st">'https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data'</span></span></span>
<span class="r-in"><span><span class="va">fastq1</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'Fastq1'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'fastq/sample1.fastq.gz'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'genome.fa'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'SIRV_isoforms_multi-fasta_170612a.fasta'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'annot.gtf'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fasta</span> <span class="op">&lt;-</span> <span class="fu"><a href="annotation_to_fasta.html">annotation_to_fasta</a></span><span class="op">(</span><span class="va">annotation</span>, <span class="va">genome_fa</span>, <span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The "phase" metadata column contains non-NA values for features of type</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   exon. This information was ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"><span><span class="fu"><a href="minimap2_realign.html">minimap2_realign</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  config <span class="op">=</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"config_sclr_nanopore_3end.json"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  fq_in <span class="op">=</span> <span class="va">fastq1</span>,</span></span>
<span class="r-in"><span>  outdir <span class="op">=</span> <span class="va">outdir</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 06:51:14 AM Fri Oct 18 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                         total mapped primary</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /tmp/RtmpZD0DsH/file21304509da85/realign2transcript.bam  4554   3764    2500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                         secondary</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /tmp/RtmpZD0DsH/file21304509da85/realign2transcript.bam      2051</span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">get_coverage</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">'realign2transcript.bam'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 103</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   transcript coverage_1 coverage_2 coverage_3 coverage_4 coverage_5 coverage_6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> SIRV101             0      0.3        0.6        0.6        0.7        0.8  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> SIRV106             0      0          0.174      0.348      0.435      0.478</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> SIRV103             0      0.143      0.571      0.786      0.786      0.786</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> SIRV105             0      0          0.625      0.844      0.906      1    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> SIRV107             0      0          0.340      0.415      0.436      0.574</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> SIRV201             0      0.182      0.455      0.455      0.455      0.455</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 96 more variables: coverage_7 &lt;dbl&gt;, coverage_8 &lt;dbl&gt;, coverage_9 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_10 &lt;dbl&gt;, coverage_11 &lt;dbl&gt;, coverage_12 &lt;dbl&gt;, coverage_13 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_14 &lt;dbl&gt;, coverage_15 &lt;dbl&gt;, coverage_16 &lt;dbl&gt;, coverage_17 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_18 &lt;dbl&gt;, coverage_19 &lt;dbl&gt;, coverage_20 &lt;dbl&gt;, coverage_21 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_22 &lt;dbl&gt;, coverage_23 &lt;dbl&gt;, coverage_24 &lt;dbl&gt;, coverage_25 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_26 &lt;dbl&gt;, coverage_27 &lt;dbl&gt;, coverage_28 &lt;dbl&gt;, coverage_29 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   coverage_30 &lt;dbl&gt;, coverage_31 &lt;dbl&gt;, coverage_32 &lt;dbl&gt;, …</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

