<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>plot read coverages — plot_coverage • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="plot read coverages — plot_coverage"><meta name="description" content="Plot the average read coverages for each length bin or a
perticular isoform"><meta property="og:description" content="Plot the average read coverages for each length bin or a
perticular isoform"><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>plot read coverages</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/model_decay.R" class="external-link"><code>R/model_decay.R</code></a></small>
      <div class="d-none name"><code>plot_coverage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot the average read coverages for each length bin or a
perticular isoform</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_coverage</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  quantiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2375</span>, <span class="fl">0.475</span>, <span class="fl">0.7125</span>, <span class="fl">0.95</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  length_bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="cn">Inf</span><span class="op">)</span>,</span>
<span>  weight_fn <span class="op">=</span> <span class="va">weight_transcripts</span>,</span>
<span>  <span class="va">filter_fn</span>,</span>
<span>  detailed <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x-">x, <a class="anchor" aria-label="anchor" href="#arg-x-"></a></dt>
<dd><p>path to the BAM file (aligning reads to the transcriptome), or
the (GenomicAlignments::readGAlignments) parsed GAlignments object, or the
tibble returned by <code><a href="get_coverage.html">get_coverage</a></code>, or the filtered tibble returned
by <code><a href="filter_coverage.html">filter_coverage</a></code>.</p></dd>


<dt id="arg-quantiles">quantiles<a class="anchor" aria-label="anchor" href="#arg-quantiles"></a></dt>
<dd><p>numeric vector to specify the quantiles to bin the transcripts lengths
by if length_bins is missing. The length bins will be determined such that the read
counts are distributed acording to the quantiles.</p></dd>


<dt id="arg-length-bins-">length_bins, <a class="anchor" aria-label="anchor" href="#arg-length-bins-"></a></dt>
<dd><p>numeric vector to specify the sizes to bin the transcripts by</p></dd>


<dt id="arg-weight-fn">weight_fn<a class="anchor" aria-label="anchor" href="#arg-weight-fn"></a></dt>
<dd><p>function to calculate the weights for the transcripts. The function
should take a numeric vector of read counts and return a numeric vector of weights.
The default function is <code><a href="weight_transcripts.html">weight_transcripts</a></code>, you can change its default
parameters by passing an anonymous function like <code>function(x) weight_transcripts(x, type = 'equal')</code>.</p></dd>


<dt id="arg-filter-fn">filter_fn<a class="anchor" aria-label="anchor" href="#arg-filter-fn"></a></dt>
<dd><p>Optional filter function to filter the transcripts before plotting.
See the <code>filter_fn</code> parameter in <code><a href="filter_coverage.html">filter_coverage</a></code> for more details.
Providing a filter fucntion here is the same as providing it in <code><a href="filter_coverage.html">filter_coverage</a></code>
and then passing the result to this function.</p></dd>


<dt id="arg-detailed">detailed<a class="anchor" aria-label="anchor" href="#arg-detailed"></a></dt>
<dd><p>logical, if <code>TRUE</code>, also plot the top 10 transcripts with the highest
read counts for each length bin.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ggplot2 object of the coverage plot(s)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create a BAM file with minimap2_realign</span></span></span>
<span class="r-in"><span><span class="va">temp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">temp_path</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">file_url</span> <span class="op">&lt;-</span> <span class="st">'https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data'</span></span></span>
<span class="r-in"><span><span class="va">fastq1</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'Fastq1'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'fastq/sample1.fastq.gz'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'genome.fa'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'SIRV_isoforms_multi-fasta_170612a.fasta'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">'annot.gtf'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">'SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'</span>, sep <span class="op">=</span> <span class="st">'/'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fasta</span> <span class="op">&lt;-</span> <span class="fu"><a href="annotation_to_fasta.html">annotation_to_fasta</a></span><span class="op">(</span><span class="va">annotation</span>, <span class="va">genome_fa</span>, <span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Import genomic features from the file as a GRanges object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Prepare the 'metadata' data frame ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Make the TxDb object ... </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The "phase" metadata column contains non-NA values for features of type</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   exon. This information was ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> OK</span>
<span class="r-in"><span><span class="fu"><a href="minimap2_realign.html">minimap2_realign</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  config <span class="op">=</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"config_sclr_nanopore_3end.json"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  fq_in <span class="op">=</span> <span class="va">fastq1</span>,</span></span>
<span class="r-in"><span>  outdir <span class="op">=</span> <span class="va">outdir</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:56:24 AM Thu Oct 17 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                         total mapped primary</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /tmp/Rtmpsdk3vz/file20fb43015280/realign2transcript.bam  4554   3764    2500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                         secondary</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /tmp/Rtmpsdk3vz/file20fb43015280/realign2transcript.bam      2051</span>
<span class="r-in"><span><span class="co"># Plot the coverages directly from the BAM file</span></span></span>
<span class="r-in"><span><span class="fu">plot_coverage</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">'realign2transcript.bam'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using quantiles to bin transcripts.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-plt img"><img src="plot_coverage-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the coverage information first</span></span></span>
<span class="r-in"><span><span class="va">coverage</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_coverage.html">get_coverage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">'realign2transcript.bam'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">read_counts</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Filter out transcripts with read counts &lt; 3</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="filter_coverage.html">filter_coverage</a></span><span class="op">(</span>filter_fn <span class="op">=</span> <span class="va">convolution_filter</span><span class="op">)</span> <span class="co"># Filter out transcripts with sharp drops / rises</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 45 transcripts found in the BAM file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 17(37.78%) transcripts failed the filter.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Failed transcripts account for 212 reads, out of 1193(17.77%) reads in total.</span>
<span class="r-in"><span><span class="co"># Plot the filtered coverages</span></span></span>
<span class="r-in"><span><span class="fu">plot_coverage</span><span class="op">(</span><span class="va">coverage</span>, detailed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using quantiles to bin transcripts.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-plt img"><img src="plot_coverage-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># filtering function can also be passed directly to plot_coverage</span></span></span>
<span class="r-in"><span><span class="fu">plot_coverage</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">'realign2transcript.bam'</span><span class="op">)</span>, filter_fn <span class="op">=</span> <span class="va">convolution_filter</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 52 transcripts found in the BAM file.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 19(36.54%) transcripts failed the filter.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Failed transcripts account for 215 reads, out of 1203(17.87%) reads in total.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using quantiles to bin transcripts.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The number of transcripts is less than the inflection index, returning equal weights for the current bin.</span>
<span class="r-plt img"><img src="plot_coverage-3.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

