<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>FLAMES Differential Transcript Usage Analysis — sc_DTU_analysis • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FLAMES Differential Transcript Usage Analysis — sc_DTU_analysis"><meta name="description" content="Chi-square based differential transcription usage analysis. This variant is meant for single cell data.
Takes the SingleCellExperiment object from sc_long_pipeline as input.
Alternatively, the path to the output folder could be provided instead of the SCE object."><meta property="og:description" content="Chi-square based differential transcription usage analysis. This variant is meant for single cell data.
Takes the SingleCellExperiment object from sc_long_pipeline as input.
Alternatively, the path to the output folder could be provided instead of the SCE object."><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>FLAMES Differential Transcript Usage Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/sc_DTU_analysis.R" class="external-link"><code>R/sc_DTU_analysis.R</code></a></small>
      <div class="d-none name"><code>sc_DTU_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Chi-square based differential transcription usage analysis. This variant is meant for single cell data.
Takes the <code>SingleCellExperiment</code> object from <code>sc_long_pipeline</code> as input.
Alternatively, the path to the output folder could be provided instead of the SCE object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sc_DTU_analysis</span><span class="op">(</span><span class="va">sce</span>, min_count <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sce">sce<a class="anchor" aria-label="anchor" href="#arg-sce"></a></dt>
<dd><p>The <code>SingleCellExperiment</code> object from <code>sc_long_pipeline</code>,
with the following metadata:
file is required under the output folder of the SCE object.</p></dd>


<dt id="arg-min-count">min_count<a class="anchor" aria-label="anchor" href="#arg-min-count"></a></dt>
<dd><p>The minimum UMI count threshold for filtering isoforms.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code>data.frame</code> containing the following columns:</p><dl><dt>gene_id</dt>
<dd><p>- differentially transcribed genes</p></dd>

 <dt>X_value</dt>
<dd><p>- the X value for the DTU gene</p></dd>

 <dt>df</dt>
<dd><p>- degrees of freedom of the approximate chi-squared distribution of the test statistic</p></dd>

 <dt>DTU_tr</dt>
<dd><p>- the transcript_id with the highest squared residuals</p></dd>

 <dt>DTU_group</dt>
<dd><p>- the cell group with the highest squared residuals</p></dd>

 <dt>p_value</dt>
<dd><p>- the p-value for the test</p></dd>

 <dt>adj_p</dt>
<dd><p>- the adjusted p-value (by Benjamini–Hochberg correction)</p></dd>


</dl><p>The table is sorted by decreasing P-values. It will also be saved as <code>sc_DTU_analysis.csv</code> under the
output folder.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function will search for genes that have at least two isoforms, each with more than <code>min_count</code> UMI counts.
For each gene, the per cell transcript counts were merged by group to generate pseudo bulk samples.
Grouping is specified by the <code>colLabels</code> of the SCE object.
The top 2 highly expressed transcripts for each group were selected and a UMI count matrix where
the rows are selected transcripts and columns are groups was used as input to a chi-square test of independence (chisq.test).
Adjusted P-values were calculated by Benjamini–Hochberg correction.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bc_allow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_allow.tsv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"rps24.fa"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bc_allow.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  destname <span class="op">=</span> <span class="va">bc_allow</span>, remove <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rps24.fa.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  destname <span class="op">=</span> <span class="va">genome_fa</span>, remove <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="sc_long_pipeline.html">sc_long_pipeline</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  genome_fa <span class="op">=</span> <span class="va">genome_fa</span>,</span></span>
<span class="r-in"><span>  fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"fastq"</span>, <span class="st">"musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rps24.gtf.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  outdir <span class="op">=</span> <span class="va">outdir</span>,</span></span>
<span class="r-in"><span>  barcodes_file <span class="op">=</span> <span class="va">bc_allow</span>,</span></span>
<span class="r-in"><span>  config_file <span class="op">=</span> <span class="fu"><a href="create_config.html">create_config</a></span><span class="op">(</span><span class="va">outdir</span>, oarfish_quantification <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Writing configuration parameters to:  /tmp/RtmpiWOPbd/file20b2454a1041/config_file_8370.json </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Configured to use 8 cores, detected 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 Start running</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 Demultiplexing using flexiplex...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matching cell barcodes...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FLEXIPLEX 0.96.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max barcode edit distance to 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max flanking sequence edit distance to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting read IDs to be  replaced</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting number of threads to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search pattern: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> primer: CTACACGACGCTCTTCCGATCT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BC: NNNNNNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UMI: NNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> polyT: TTTTTTTTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting known barcodes from /tmp/RtmpiWOPbd/file20b2454a1041/bc_allow.tsv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of known barcodes: 143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Processing file: /__w/_temp/Library/FLAMES/extdata/fastq/musc_rps24.fastq.gz</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Searching for barcodes...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads processed: 393</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where a barcode was found: 368</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where more than one barcode was found: 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Skipping TSO trimming...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 Demultiplex done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running FLAMES pipeline...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Input parameters:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "pipeline_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "seed": [2022],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "threads": [8],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_barcode_demultiplex": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_gene_quantification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_genome_alignment": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_isoform_identification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "bambu_isoform_identification": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "multithread_isoform_identification": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_read_realignment": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_transcript_quantification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "oarfish_quantification": [false]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "barcode_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_bc_editdistance": [2],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_flank_editdistance": [8],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "pattern": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "primer": ["CTACACGACGCTCTTCCGATCT"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "BC": ["NNNNNNNNNNNNNNNN"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "UMI": ["NNNNNNNNNNNN"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "polyT": ["TTTTTTTTT"]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "TSO_seq": ["CCCATGTACTCTGCGTTGATACCACTGCTT"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "TSO_prime": [3],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "full_length_only": [false]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "isoform_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "generate_raw_isoform": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_dist": [10],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_ts_dist": [100],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_splice_match_dist": [10],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_fl_exon_len": [40],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_site_per_splice": [3],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_sup_cnt": [5],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_cnt_pct": [0.001],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_sup_pct": [0.2],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "bambu_trust_reference": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "strand_specific": [0],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "remove_incomp_reads": [4],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "downsample_ratio": [1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "alignment_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "use_junctions": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "no_flank": [false]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "realign_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "use_annotation": [true]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "transcript_counting": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_tr_coverage": [0.4],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_read_coverage": [0.4]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> } </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene annotation: /__w/_temp/Library/FLAMES/extdata/rps24.gtf.gz </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genome fasta: /tmp/RtmpiWOPbd/file20b2454a1041/rps24.fa </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> input fastq: /tmp/RtmpiWOPbd/file20b2454a1041/matched_reads.fastq </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> output directory: /tmp/RtmpiWOPbd/file20b2454a1041 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> minimap2 path: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> k8 path: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Aligning reads to genome using minimap2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 minimap2_align</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 Start gene quantification and UMI deduplication</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:58 AM Mon Oct 07 2024 quantify genes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Found genome alignment file(s): 	align2genome.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 Gene quantification and UMI deduplication done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 Start isoform identificaiton</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 find_isoform</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 Isoform identificaiton done</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Realigning deduplicated reads to transcript using minimap2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Generating transcript count matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03:24:59 AM Mon Oct 07 2024 quantify transcripts </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Found realignment file(s): 	realign2transcript.bam</span>
<span class="r-in"><span><span class="va">group_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>barcode_seq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, groups <span class="op">=</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="st">"ENSMUST00000169826.2"</span>, <span class="op">]</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">group_anno</span><span class="op">$</span><span class="va">groups</span></span></span>
<span class="r-in"><span><span class="fu">sc_DTU_analysis</span><span class="op">(</span><span class="va">sce</span>, min_count <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loading isoform_FSM_annotation.csv ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Selecting transcript_ids with full splice match ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Summing transcripts with same FSM ... </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Creating FSM_count.csv.gz ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /tmp/RtmpiWOPbd/file20b2454a1041/FSM_count.csv.gz saved.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	 6 FSM_match(s) found.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Filtering for genes with at least 2 detected isforms ...      6 FSM_match(s) left.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Keeping only the top 4 expressed FSM_matches for each gene ...      4 FSM_match(s) left.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Aggregating counts by cluster labels ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Filtering isoforms ... 4  transcript_id(s) remaining.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Performing Chi-square tests ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Chi-squared approximation(s) may be incorrect</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Results saved to  /tmp/RtmpiWOPbd/file20b2454a1041/sc_DTU_analysis.csv </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         gene_id  X_value df                          DTU_tr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X-squared ENSMUSG00000025290.17 16.72638  3 ENSMUSG00000025290.17_19_5159_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           DTU_group      p_value        adj_p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X-squared     FALSE 0.0008044732 0.0008044732</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

