<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>FLAMES heetmap plots — sc_heatmap_expression • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FLAMES heetmap plots — sc_heatmap_expression"><meta name="description" content="Plot expression heatmap of top n isoforms of a gene"><meta property="og:description" content="Plot expression heatmap of top n isoforms of a gene"><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>FLAMES heetmap plots</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/sc_annotate_plots.R" class="external-link"><code>R/sc_annotate_plots.R</code></a></small>
      <div class="d-none name"><code>sc_heatmap_expression.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot expression heatmap of top n isoforms of a gene</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sc_heatmap_expression</span><span class="op">(</span></span>
<span>  <span class="va">gene</span>,</span>
<span>  <span class="va">multiAssay</span>,</span>
<span>  impute <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_isoforms <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="va">transcript_ids</span>,</span>
<span>  n_pcs <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  isoform_legend_width <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  col_low <span class="op">=</span> <span class="st">"#313695"</span>,</span>
<span>  col_mid <span class="op">=</span> <span class="st">"#FFFFBF"</span>,</span>
<span>  col_high <span class="op">=</span> <span class="st">"#A50026"</span>,</span>
<span>  color_quantile <span class="op">=</span> <span class="fl">0.95</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gene">gene<a class="anchor" aria-label="anchor" href="#arg-gene"></a></dt>
<dd><p>The gene symbol of interest.</p></dd>


<dt id="arg-multiassay">multiAssay<a class="anchor" aria-label="anchor" href="#arg-multiassay"></a></dt>
<dd><p>The <code>MultiAssayExperiment</code> object from <code><a href="combine_sce.html">combine_sce()</a></code>.</p></dd>


<dt id="arg-impute">impute<a class="anchor" aria-label="anchor" href="#arg-impute"></a></dt>
<dd><p>Whether to impute expression levels for cells without transcript counts</p></dd>


<dt id="arg-n-isoforms">n_isoforms<a class="anchor" aria-label="anchor" href="#arg-n-isoforms"></a></dt>
<dd><p>The number of expressed isoforms to keep.</p></dd>


<dt id="arg-transcript-ids">transcript_ids<a class="anchor" aria-label="anchor" href="#arg-transcript-ids"></a></dt>
<dd><p>specify the transcript ids instead of selecting the top <code>n_isoforms</code></p></dd>


<dt id="arg-n-pcs">n_pcs<a class="anchor" aria-label="anchor" href="#arg-n-pcs"></a></dt>
<dd><p>The number of principal components to generate.</p></dd>


<dt id="arg-isoform-legend-width">isoform_legend_width<a class="anchor" aria-label="anchor" href="#arg-isoform-legend-width"></a></dt>
<dd><p>The width of isoform legends in heatmaps, in <code>cm</code>.</p></dd>


<dt id="arg-col-low">col_low<a class="anchor" aria-label="anchor" href="#arg-col-low"></a></dt>
<dd><p>Color for cells with low expression levels in UMAPs.</p></dd>


<dt id="arg-col-mid">col_mid<a class="anchor" aria-label="anchor" href="#arg-col-mid"></a></dt>
<dd><p>Color for cells with intermediate expression levels in UMAPs.</p></dd>


<dt id="arg-col-high">col_high<a class="anchor" aria-label="anchor" href="#arg-col-high"></a></dt>
<dd><p>Color for cells with high expression levels in UMAPs.</p></dd>


<dt id="arg-color-quantile">color_quantile<a class="anchor" aria-label="anchor" href="#arg-color-quantile"></a></dt>
<dd><p>The lower and upper expression quantile to be displayed bewteen <code>col_low</code> and <code>col_high</code>, e.g. with <code>color_quantile = 0.95</code>, cells with expressions higher than 95% of other cells will all be shown in <code>col_high</code>, and cells with expression lower than 95% of other cells will all be shown in <code>col_low</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code>ggplot</code> object of the heatmap</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes the combined <code>MultiAssayExperiment</code> object from
<code>combine_sce</code> and plots an expression heatmap with the isoform alignment visualisations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">combined_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="combine_sce.html">combine_sce</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    short_read_large <span class="op">=</span> <span class="va">scmixology_lib90</span>,</span></span>
<span class="r-in"><span>    short_read_small <span class="op">=</span> <span class="va">scmixology_lib10</span>,</span></span>
<span class="r-in"><span>    long_read_sce <span class="op">=</span> <span class="va">scmixology_lib10_transcripts</span>,</span></span>
<span class="r-in"><span>    remove_duplicates <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>sampleMap[['assay']] coerced with as.factor()</span>
<span class="r-in"><span><span class="fu">sc_heatmap_expression</span><span class="op">(</span>gene <span class="op">=</span> <span class="st">"ENSG00000108107"</span>, multiAssay <span class="op">=</span> <span class="va">combined_sce</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running PCA for experiments(multiAssay)$gene_counts ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> harmonizing input:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 257 sampleMap rows with 'colname' not in colnames of experiments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 236 colData rownames not in sampleMap 'primary'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running UMAP for experiments(multiAssay)$gene_counts ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>'experiments' dropped; see 'drops()'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> harmonizing input:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 999 sampleMap rows not in names(experiments)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 21 colData rownames not in sampleMap 'primary'</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>'experiments' dropped; see 'drops()'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> harmonizing input:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 999 sampleMap rows not in names(experiments)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 21 colData rownames not in sampleMap 'primary'</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>'experiments' dropped; see 'drops()'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> harmonizing input:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 4743 sampleMap rows not in names(experiments)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   removing 3765 colData rownames not in sampleMap 'primary'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Constructing graphics...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `group.selfish`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Ignoring unknown parameters: `label` and `utr.geom`</span>
<span class="r-plt img"><img src="sc_heatmap_expression-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

