<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pipeline for Multi-sample Single Cell Data — sc_long_multisample_pipeline • FLAMES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pipeline for Multi-sample Single Cell Data — sc_long_multisample_pipeline"><meta name="description" content="Semi-supervised isoform detection and annotation for long read data.
This variant is for multi-sample single cell data. By default, this pipeline demultiplexes input
fastq data (match_cell_barcode = TRUE). Specific parameters relating to
analysis can be changed either through function arguments, or through a
configuration JSON file."><meta property="og:description" content="Semi-supervised isoform detection and annotation for long read data.
This variant is for multi-sample single cell data. By default, this pipeline demultiplexes input
fastq data (match_cell_barcode = TRUE). Specific parameters relating to
analysis can be changed either through function arguments, or through a
configuration JSON file."><meta property="og:image" content="https://mritchielab.github.io/FLAMES/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Pipeline for Multi-sample Single Cell Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/R/sc_long_multisample_pipeline.R" class="external-link"><code>R/sc_long_multisample_pipeline.R</code></a></small>
      <div class="d-none name"><code>sc_long_multisample_pipeline.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Semi-supervised isoform detection and annotation for long read data.
This variant is for multi-sample single cell data. By default, this pipeline demultiplexes input
fastq data (<code>match_cell_barcode = TRUE</code>). Specific parameters relating to
analysis can be changed either through function arguments, or through a
configuration JSON file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sc_long_multisample_pipeline</span><span class="op">(</span></span>
<span>  <span class="va">annotation</span>,</span>
<span>  <span class="va">fastqs</span>,</span>
<span>  <span class="va">outdir</span>,</span>
<span>  <span class="va">genome_fa</span>,</span>
<span>  sample_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimap2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k8 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  barcodes_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expect_cell_numbers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  config_file <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-annotation">annotation<a class="anchor" aria-label="anchor" href="#arg-annotation"></a></dt>
<dd><p>The file path to the annotation file in GFF3 format</p></dd>


<dt id="arg-fastqs">fastqs<a class="anchor" aria-label="anchor" href="#arg-fastqs"></a></dt>
<dd><p>The input fastq files for multiple samples. It can be provided in different way: 1) a single path to
the folder containing fastq files, each fastq file will be treated as a sample; or 2) a vector of
paths to each fastq file, each fastq file will be treated as a sample; or 3) a vector of paths to
folders containing fastq files, each folder will be treated as a sample.</p></dd>


<dt id="arg-outdir">outdir<a class="anchor" aria-label="anchor" href="#arg-outdir"></a></dt>
<dd><p>The path to directory to store all output files.</p></dd>


<dt id="arg-genome-fa">genome_fa<a class="anchor" aria-label="anchor" href="#arg-genome-fa"></a></dt>
<dd><p>The file path to genome fasta file.</p></dd>


<dt id="arg-sample-names">sample_names<a class="anchor" aria-label="anchor" href="#arg-sample-names"></a></dt>
<dd><p>A vector of sample names,
Default to the file names of input fastq files, or folder names if <code>fastqs</code> is
a vector of folders.</p></dd>


<dt id="arg-minimap-">minimap2<a class="anchor" aria-label="anchor" href="#arg-minimap-"></a></dt>
<dd><p>Path to minimap2, if it is not in PATH. Only required if either or both of
<code>do_genome_align</code> and <code>do_read_realign</code> are <code>TRUE</code>.</p></dd>


<dt id="arg-k-">k8<a class="anchor" aria-label="anchor" href="#arg-k-"></a></dt>
<dd><p>Path to the k8 Javascript shell binary. Only required if <code>do_genome_align</code> is <code>TRUE</code>.</p></dd>


<dt id="arg-barcodes-file">barcodes_file<a class="anchor" aria-label="anchor" href="#arg-barcodes-file"></a></dt>
<dd><p>The file path to the reference csv used for demultiplexing in flexiplex. If not specified, the
demultiplexing will be performed using BLAZE. Default is <code>NULL</code>.</p></dd>


<dt id="arg-expect-cell-numbers">expect_cell_numbers<a class="anchor" aria-label="anchor" href="#arg-expect-cell-numbers"></a></dt>
<dd><p>A vector of roughly expected numbers of cells in each sample E.g., the targeted number of cells.
Required if using BLAZE for demultiplexing, specifically, when the <code>do_barcode_demultiplex</code>
are <code>TRUE</code> in the the JSON configuration file and <code>barcodes_file</code> is not specified.
Default is <code>NULL</code>.</p></dd>


<dt id="arg-config-file">config_file<a class="anchor" aria-label="anchor" href="#arg-config-file"></a></dt>
<dd><p>File path to the JSON configuration file. If specified, <code>config_file</code> overrides
all configuration parameters</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list of <code>SingleCellExperiment</code> objects if "do_transcript_quantification" set to true.
Otherwise nothing will be returned.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>By default FLAMES use minimap2 for read alignment. After the genome alignment step (<code>do_genome_align</code>), FLAMES summarizes the alignment for each read in every sample by grouping reads
with similar splice junctions to get a raw isoform annotation (<code>do_isoform_id</code>). The raw isoform
annotation is compared against the reference annotation to correct potential splice site
and transcript start/end errors. Transcripts that have similar splice junctions
and transcript start/end to the reference transcript are merged with the
reference. This process will also collapse isoforms that are likely to be truncated
transcripts. If <code>isoform_id_bambu</code> is set to <code>TRUE</code>, <code><a href="https://rdrr.io/pkg/bambu/man/bambu.html" class="external-link">bambu::bambu</a></code> will be used to generate the updated annotations (Not implemented for multi-sample yet).
Next is the read realignment step (<code>do_read_realign</code>), where the sequence of each transcript from the update annotation is extracted, and
the reads are realigned to this updated <code>transcript_assembly.fa</code> by minimap2. The
transcripts with only a few full-length aligned reads are discarded (Not implemented for multi-sample yet).
The reads are assigned to transcripts based on both alignment score, fractions of
reads aligned and transcript coverage. Reads that cannot be uniquely assigned to
transcripts or have low transcript coverage are discarded. The UMI transcript
count matrix is generated by collapsing the reads with the same UMI in a similar
way to what is done for short-read scRNA-seq data, but allowing for an edit distance
of up to 2 by default. Most of the parameters, such as the minimal distance to splice site and minimal percentage of transcript coverage
can be modified by the JSON configuration file (<code>config_file</code>).</p>
<p>The default parameters can be changed either through the function
arguments are through the configuration JSON file <code>config_file</code>. the <code>pipeline_parameters</code>
section specifies which steps are to be executed in the pipeline - by default, all
steps are executed. The <code>isoform_parameters</code> section affects isoform detection - key
parameters include:</p><dl><dt><code>Min_sup_cnt</code></dt>
<dd><p>which causes transcripts with less reads aligned than
it's value to be discarded</p></dd>

<dt><code>MAX_TS_DIST</code></dt>
<dd><p>which merges transcripts with the same intron
chain and TSS/TES distace less than <code>MAX_TS_DIST</code></p></dd>

<dt><code>strand_specific</code></dt>
<dd><p>which specifies if reads are in the same strand as the mRNA (1),
or the reverse complemented (-1) or not strand specific (0), which results in
strand information being based on reference annotation.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="bulk_long_pipeline.html">bulk_long_pipeline()</a></code> for bulk long data,
<code>SingleCellExperiment()</code> for how data is outputted</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu">ShortRead</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ShortRead/man/readFastq.html" class="external-link">readFastq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fastq/musc_rps24.fastq.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bc_allow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"bc_allow.tsv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"rps24.fa"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bc_allow.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>, destname <span class="op">=</span> <span class="va">bc_allow</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://henrikbengtsson.github.io/R.utils/reference/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.fa.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>, destname <span class="op">=</span> <span class="va">genome_fa</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ShortRead</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ShortRead/man/readFastq.html" class="external-link">writeFastq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">reads</span>, size <span class="op">=</span> <span class="fl">500</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq/sample1.fq.gz"</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"w"</span>, full <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ShortRead</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ShortRead/man/readFastq.html" class="external-link">writeFastq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">reads</span>, size <span class="op">=</span> <span class="fl">500</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq/sample2.fq.gz"</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"w"</span>, full <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ShortRead</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ShortRead/man/readFastq.html" class="external-link">writeFastq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">reads</span>, size <span class="op">=</span> <span class="fl">500</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq/sample3.fq.gz"</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"w"</span>, full <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="find_bin.html">find_bin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">sce_list</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu">sc_long_multisample_pipeline</span><span class="op">(</span></span></span>
<span class="r-in"><span>    annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/rps24.gtf.gz"</span>, package <span class="op">=</span> <span class="st">"FLAMES"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    fastqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"fastq"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    outdir <span class="op">=</span> <span class="va">outdir</span>,</span></span>
<span class="r-in"><span>    genome_fa <span class="op">=</span> <span class="va">genome_fa</span>,</span></span>
<span class="r-in"><span>    barcodes_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">bc_allow</span>, <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No config file provided, creating a default config in /tmp/Rtmphcds2N/file1c01433f87fc </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Writing configuration parameters to:  /tmp/Rtmphcds2N/file1c01433f87fc/config_file_7169.json </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Configured to use 8 cores, detected 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FLEXIPLEX 0.96.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max barcode edit distance to 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max flanking sequence edit distance to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting read IDs to be  replaced</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting number of threads to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search pattern: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> primer: CTACACGACGCTCTTCCGATCT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BC: NNNNNNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UMI: NNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> polyT: TTTTTTTTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting known barcodes from /tmp/Rtmphcds2N/file1c01433f87fc/bc_allow.tsv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of known barcodes: 143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Searching for barcodes...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads processed: 500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where a barcode was found: 459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where more than one barcode was found: 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Skipping TSO trimming...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FLEXIPLEX 0.96.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max barcode edit distance to 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max flanking sequence edit distance to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting read IDs to be  replaced</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting number of threads to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search pattern: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> primer: CTACACGACGCTCTTCCGATCT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BC: NNNNNNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UMI: NNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> polyT: TTTTTTTTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting known barcodes from /tmp/Rtmphcds2N/file1c01433f87fc/bc_allow.tsv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of known barcodes: 143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Searching for barcodes...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads processed: 500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where a barcode was found: 465</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where more than one barcode was found: 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Skipping TSO trimming...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FLEXIPLEX 0.96.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max barcode edit distance to 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting max flanking sequence edit distance to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting read IDs to be  replaced</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting number of threads to 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search pattern: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> primer: CTACACGACGCTCTTCCGATCT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BC: NNNNNNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UMI: NNNNNNNNNNNN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> polyT: TTTTTTTTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Setting known barcodes from /tmp/Rtmphcds2N/file1c01433f87fc/bc_allow.tsv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of known barcodes: 143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Searching for barcodes...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads processed: 500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where a barcode was found: 469</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads where more than one barcode was found: 12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Skipping TSO trimming...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running FLAMES pipeline...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Input parameters:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "comment": ["this is the default config for nanopore single cell long read data using 10x 3'end kit. use splice annotation in alignment. "],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "pipeline_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "seed": [2022],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "threads": [8],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_barcode_demultiplex": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_gene_quantification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_genome_alignment": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_isoform_identification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "bambu_isoform_identification": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "multithread_isoform_identification": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_read_realignment": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "do_transcript_quantification": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "oarfish_quantification": [true]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "barcode_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_bc_editdistance": [2],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_flank_editdistance": [8],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "pattern": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "primer": ["CTACACGACGCTCTTCCGATCT"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "BC": ["NNNNNNNNNNNNNNNN"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "UMI": ["NNNNNNNNNNNN"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       "polyT": ["TTTTTTTTT"]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "TSO_seq": ["CCCATGTACTCTGCGTTGATACCACTGCTT"],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "TSO_prime": [3],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "full_length_only": [false]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "isoform_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "generate_raw_isoform": [false],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_dist": [10],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_ts_dist": [100],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_splice_match_dist": [10],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_fl_exon_len": [40],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "max_site_per_splice": [3],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_sup_cnt": [5],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_cnt_pct": [0.001],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_sup_pct": [0.2],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "bambu_trust_reference": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "strand_specific": [0],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "remove_incomp_reads": [4],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "downsample_ratio": [1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "alignment_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "use_junctions": [true],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "no_flank": [false]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "realign_parameters": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "use_annotation": [true]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   },</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   "transcript_counting": {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_tr_coverage": [0.4],</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     "min_read_coverage": [0.4]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }gene annotation: /__w/_temp/Library/FLAMES/extdata/rps24.gtf.gz </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genome fasta: /tmp/Rtmphcds2N/file1c01433f87fc/rps24.fa </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> input fastqs: /tmp/Rtmphcds2N/file1c01433f87fc/sample1_matched_reads.fastq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  /tmp/Rtmphcds2N/file1c01433f87fc/sample2_matched_reads.fastq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  /tmp/Rtmphcds2N/file1c01433f87fc/sample3_matched_reads.fastq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> output directory: /tmp/Rtmphcds2N/file1c01433f87fc </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> minimap2 path: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> k8 path: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Aligning reads to genome using minimap2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Aligning sample  sample1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:51 AM Mon Sep 30 2024 minimap2_align</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Aligning sample  sample2 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:51 AM Mon Sep 30 2024 minimap2_align</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Aligning sample  sample3 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:51 AM Mon Sep 30 2024 minimap2_align</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:51 AM Mon Sep 30 2024 quantify genes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Found genome alignment file(s): 	sample1_align2genome.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	sample2_align2genome.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	sample3_align2genome.bam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:52 AM Mon Sep 30 2024 find_isoform</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Realign to transcript using minimap2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Realigning sample  sample1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:52 AM Mon Sep 30 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by  CB </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Realigning sample  sample2 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:53 AM Mon Sep 30 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by  CB </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Realigning sample  sample3 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05:56:54 AM Mon Sep 30 2024 minimap2_realign</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sorting by  CB </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> #### Generating transcript count matrix</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

