<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>FLAMES • FLAMES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FLAMES">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLAMES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.11.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/FLAMES_vignette.html">FLAMES package demo</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mritchielab/FLAMES"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://bioconductor.org/packages/FLAMES"><span class="fa fa-book"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>FLAMES</h1>
                        <h4 data-toc-skip class="author">Oliver Voogd,
Changqing Wang, Yupei You</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/FLAMES/blob/devel/vignettes/FLAMES_vignette.Rmd" class="external-link"><code>vignettes/FLAMES_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>FLAMES_vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="flames">FLAMES<a class="anchor" aria-label="anchor" href="#flames"></a>
</h2>
<p>The FLAMES package provides a framework for performing single-cell
and bulk read full-length analysis of mutations and splicing. FLAMES
performs cell barcode and UMI assignment from nanopore reads as well as
semi-supervised isoform detection and quantification. FLAMES is designed
to be an easy and quick to use, powerful workflow for isoform detection
and quantification, splicing analysis and mutation detection, and seeks
to overcome the limitations of other tools, such as an inability to
process single cell data, and a focus on cell barcode and UMI assignment
<span class="citation">(Tian et al. 2020)</span>.</p>
<p>This R package represents an enhanced iteration of the FLAMES Python
module that originally designed to support the research work presented
in <em>Comprehensive characterization of single-cell full-length
isoforms in human and mouse with long-read sequencing</em> by Tian et al
<span class="citation">(2020)</span>. This upgraded R version not only
simplifies the installation and execution processes but also
incorporates additional functionality, streamlining the analysis of
single-cell full-length isoforms using long-read sequencing data.</p>
<div class="float">
<img src="FLAMESpipeline-01.png" width="800" alt="(#fig:workflow) FLAMES pipeline"><div class="figcaption">(#fig:workflow) FLAMES pipeline</div>
</div>
<p>When processing single cell data, FLAMES can be run on data generated
from long-read platform with or without matched short-read sequencing
data. If only the long reads are available, FLAMES takes as input the
long reads in fastq files. The <code><a href="../reference/blaze.html">blaze()</a></code> function is then
called to demultiplex the long reads into cell-specific fastq files. If
the short-read data are unavailable, FLAMES incorporates the <a href="https://github.com/shimlab/BLAZE" class="external-link">BLAZE</a> as
<code><a href="../reference/blaze.html">blaze()</a></code> function to identify cell barcodes solely from long
reads <span class="citation">(You et al. 2023)</span>. The
<code><a href="../reference/blaze.html">blaze()</a></code> function is called to locate the barcode sequencing
in the long-reads, identify barcode allow-list directly from long reads,
assign reads to cells (i.e., demultiplexing) and trims the cell barcodes
and the flanking UMI sequence.</p>
<p>When matched short-read single-cell RNA sequencing data is available,
FLAMES could take the cell barcode allow-list generated from short reads
in addition to the long-read fastq files. The short-read allow-list will
used as a reference to guide the demultiplexing of the long reads. To do
this, FLAMES incorporates the <a href="https://davidsongroup.github.io/flexiplex/" class="external-link">flexiplex</a> as
<code><a href="../reference/find_barcode.html">find_barcode()</a></code> function to perform demultiplexing <span class="citation">(Davidson et al. 2023)</span>. The
<code><a href="../reference/find_barcode.html">find_barcode()</a></code> function is called to extract the cell
barcode and UMI sequence from the long reads, using the allow-list as a
reference.</p>
<p>After the demultiplexing, the pipeline calls
<code>minimap_align()</code>, a <code>minimap2</code> wrapper, to align
the demultiplex long-reads to the reference genome.
<code><a href="../reference/quantify_gene.html">quantify_gene()</a></code> is then called to deduplicate reads from
same unique molecular identifiers (UMIs) and generate gene-level UMI
counts. Note that during the UMI deduplication, the pipeline only keeps
the longest reads among those with the same UMI for downstream
steps.</p>
<p>Next, <code><a href="../reference/find_isoform.html">find_isoform()</a></code> is called to identify novel
isoforms using the alignment, creating an updated transcript assembly.
In this step, the users may choose to use <a href="https://github.com/GoekeLab/bambu" class="external-link">bambu</a>, which is designed
for transcript discovery and quantification using long read RNA-Seq data
<span class="citation">(Chen et al. 2023)</span>. Otherwise, users could
use the build-in isoform identification methods in FLAMES. Both methods
have been wrapped in the <code><a href="../reference/find_isoform.html">find_isoform()</a></code> function.</p>
<p>Afterwards, the demultiplexed reads are aligned to the transcript
assembly by the function <code>minimap_realign()</code>. Finally, FLAMES
calls <code><a href="../reference/quantify_transcript.html">quantify_transcript()</a></code> to quantify the transcript
counts using the (re-)alignment file.</p>
<p>Figure @ref(fig:workflow) summerise the steps of the pipeline
described above. The pipeline functions (<code>sc_long_pipeline</code>,
<code>bulk_long_pipeline</code>,
<code>sc_long_multisample_pipline</code>) execute the steps sequentially
and return <code>SingleCellExperiment</code> object,
<code>SummarizedExperiment</code> object and a list of
<code>SingleCellExperiment</code> objects respectivly.</p>
<p>For read alignment and realignment, FLAMES uses <a href="https://github.com/lh3/minimap2" class="external-link">minimap2</a>, a versatile
alignment program for aligning sequences against a reference database
<span class="citation">(Li 2018)</span>. This software needs to be
downloaded prior to using the FLAMES pipeline, and can be found at <a href="https://github.com/lh3/minimap2" class="external-link">https://github.com/lh3/minimap2</a>.</p>
</div>
<div class="section level2">
<h2 id="flames-pipeline-execution">FLAMES Pipeline Execution<a class="anchor" aria-label="anchor" href="#flames-pipeline-execution"></a>
</h2>
<p>This vignette will detail the process of running the FLAMES pipeline.
It details the execution of both the single cell pipeline
(<code><a href="../reference/sc_long_pipeline.html">sc_long_pipeline()</a></code>) and the bulk data pipeline
(<code><a href="../reference/bulk_long_pipeline.html">bulk_long_pipeline()</a></code>).</p>
<div class="section level3">
<h3 id="flames-single-cell-pipeline">FLAMES Single Cell Pipeline<a class="anchor" aria-label="anchor" href="#flames-single-cell-pipeline"></a>
</h3>
<div class="section level4">
<h4 id="environment-setup">Environment setup<a class="anchor" aria-label="anchor" href="#environment-setup"></a>
</h4>
<p>To get started, the pipeline needs access to a gene annotation file
in GFF3 or GTF format, a directory containing one or more FASTQ files
(which will be merged as pre-processing), a genome FASTA file, as well
as the file path to minimap2, and the file path to the directory to hold
output files.</p>
<p>The single cell pipeline can demultiplex the input data before
running, if required. This can be disabled by setting the
<code>do_barcode_demultiplex</code> argument in the config file to
<code>FALSE</code> when calling the pipeline. This example dataset has
already been demultiplexed.</p>
<p>For this example, the required files are downloaded from GitHub using
<a href="http://bioconductor.org/packages/release/bioc/html/BiocFileCache.html" class="external-link">BiocFileCache</a>
<span class="citation">(Shepherd and Morgan 2021)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">temp_path</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">file_url</span> <span class="op">&lt;-</span></span>
<span>  <span class="st">"https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data"</span></span>
<span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>  <span class="va">bfc</span>, <span class="st">"Annotation"</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"gencodeshortened.gtf"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>  <span class="va">bfc</span>,</span>
<span>  <span class="st">"Genomefa"</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"GRCh38shortened.fa"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">fastq</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>  <span class="va">bfc</span>, <span class="st">"Fastq"</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"sc_align2genome.sample.fastq.gz"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># setup other environment variables</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span><span class="va">config_file</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">outdir</span>, type <span class="op">=</span> <span class="st">"SIRV"</span>, do_barcode_demultiplex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">#&gt;   method from   </span></span>
<span><span class="co">#&gt;   +.gg   ggplot2</span></span>
<span><span class="co">#&gt; Writing configuration parameters to:  /tmp/RtmpsYpfbT/file1d782cd215af/config_file_7544.json</span></span></code></pre></div>
<p>The optional argument <code>config_file</code> can be given to both
<code><a href="../reference/bulk_long_pipeline.html">bulk_long_pipeline()</a></code> and <code><a href="../reference/sc_long_pipeline.html">sc_long_pipeline()</a></code> in
order to customise the execution of the pipelines. It is expected to be
a JSON file, and an example can be found by calling
<code><a href="../reference/create_config.html">FLAMES::create_config</a></code>, which returns the path to a copy of
the default JSON configuration file. The values from the default configs
can be altered by either editing the JSON file manually, or passing
additional arguments to <code><a href="../reference/create_config.html">FLAMES::create_config</a></code>, for example,
<code>FLAMES::create_config(outdir, type = "sc_3end", min_sup_cnt = 10)</code>
with create a copy of the default config for 10X 3’ end nanopore single
cell reads, with the <code>min_sup_cnt</code> value (minimal number of
supporting reads for a novel isoform to pass filtering) changed to
10.</p>
<p>This vignette uses the default configuration file created for SIRV
reads.</p>
</div>
<div class="section level4">
<h4 id="flames-execution">FLAMES execution<a class="anchor" aria-label="anchor" href="#flames-execution"></a>
</h4>
<p>Once the initial variables have been setup, the pipeline can be run
using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mritchielab.github.io/FLAMES">FLAMES</a></span><span class="op">)</span></span>
<span><span class="co"># do not run if minimap2 cannot be found</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="../reference/sys_which.html">sys_which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sc_long_pipeline.html">sc_long_pipeline</a></span><span class="op">(</span></span>
<span>    annotation <span class="op">=</span> <span class="va">annot</span>, fastq <span class="op">=</span> <span class="va">fastq</span>, genome_fa <span class="op">=</span> <span class="va">genome_fa</span>,</span>
<span>    outdir <span class="op">=</span> <span class="va">outdir</span>, config_file <span class="op">=</span> <span class="va">config_file</span>, expect_cell_number <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>If, however, the input fastq files need to be demultiplexed, then the
<code>reference_csv</code> argument will need to be specified -
<code>reference_csv</code> is the file path to a cell barcode
allow-list, as a text file with one barcode per line. The
<code>filtered_feature_bc_matrix/barcodes.tsv.gz</code> from
<code>cellranger</code> outputs can be used to create such allow-list,
with
<code>zcat barcodes.tsv.gz | cut -f1 -d'-' &gt; allow-list.csv</code>.</p>
<p>The pipeline can alse be run by calling the consituent steps
sequentially:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mritchielab.github.io/FLAMES">FLAMES</a></span><span class="op">)</span></span>
<span><span class="co"># do not run if minimap2 cannot be found</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="../reference/sys_which.html">sys_which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">config</span> <span class="op">&lt;-</span> <span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span><span class="va">config_file</span><span class="op">)</span></span>
<span>  <span class="co"># find_barcode(...)</span></span>
<span>  <span class="va">genome_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="../reference/minimap2_align.html">minimap2_align</a></span><span class="op">(</span></span>
<span>    config <span class="op">=</span> <span class="va">config</span>, fa_file <span class="op">=</span> <span class="va">genome_fa</span>, fq_in <span class="op">=</span> <span class="va">fastq</span>, annot <span class="op">=</span> <span class="va">annot</span>,</span>
<span>    outdir <span class="op">=</span> <span class="va">outdir</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/find_isoform.html">find_isoform</a></span><span class="op">(</span></span>
<span>    annotation <span class="op">=</span> <span class="va">annot</span>, genome_fa <span class="op">=</span> <span class="va">genome_fa</span>,</span>
<span>    genome_bam <span class="op">=</span> <span class="va">genome_bam</span>, outdir <span class="op">=</span> <span class="va">outdir</span>, config <span class="op">=</span> <span class="va">config</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/minimap2_realign.html">minimap2_realign</a></span><span class="op">(</span></span>
<span>    config <span class="op">=</span> <span class="va">config</span>, fq_in <span class="op">=</span> <span class="va">fastq</span>,</span>
<span>    outdir <span class="op">=</span> <span class="va">outdir</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/quantify_transcript.html">quantify_transcript</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="va">annot</span>, outdir <span class="op">=</span> <span class="va">outdir</span>, config <span class="op">=</span> <span class="va">config</span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_sce_from_dir.html">create_sce_from_dir</a></span><span class="op">(</span>outdir <span class="op">=</span> <span class="va">outdir</span>, annotation <span class="op">=</span> <span class="va">annot</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visulisation">Visulisation<a class="anchor" aria-label="anchor" href="#visulisation"></a>
</h3>
<p>The <code><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim()</a></code> function can be used to
visualise the isoform expression in reduced dimensions. The function
takes a <code>SingleCellExperiment</code> object and either a gene ID or
a vector of transcript IDs as input. If the
<code>SingleCellExperiment</code> object contains gene counts with the
transcript counts stored as the “transcript” <code>altExp</code> slot,
the function will use the reduced dimensions of the gene counts and plot
the transcript expression values.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mritchielab.github.io/FLAMES">FLAMES</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># just the transcript counts</span></span>
<span><span class="va">scmixology_lib10_transcripts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim</a></span><span class="op">(</span><span class="st">'ENSG00000108107'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMES_vignette_files/figure-html/unnamed-chunk-4-1.png" width="540"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># SCE with gene counts as main assay and transcript counts as altExp</span></span>
<span><span class="va">scmixology_lib10</span> <span class="op">&lt;-</span> <span class="va">scmixology_lib10</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">scmixology_lib10</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">sce_lr</span> <span class="op">&lt;-</span> <span class="va">scmixology_lib10</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scmixology_lib10</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scmixology_lib10_transcripts</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce_lr</span>, <span class="st">"transcript"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scmixology_lib10_transcripts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce_lr</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sce_lr</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim</a></span><span class="op">(</span><span class="st">'ENSG00000108107'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMES_vignette_files/figure-html/unnamed-chunk-4-2.png" width="540"></p>
<div class="section level4">
<h4 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h4>
<p>For experiments where the long-read is preformed on a subsample of
cells and short-reads are preformed on the full sample, the
<code><a href="../reference/combine_sce.html">combine_sce()</a></code> function can be used to combine the
<code>SingleCellExperiment</code> objects of the two samples. The first
argument should be the subsample where the main assay is the gene counts
and the transcript counts is saved as the “transcript”
<code>altExp</code> slot. The second argument should be the other sample
with only gene counts. This function returns a
<code>SingleCellExperiment</code> object with the main assay containing
the combined gene counts and a “transcript” <code>altExp</code> slot,
where cells missing long-read data have <code>NA</code> values. The
<code><a href="../reference/sc_impute_transcript.html">sc_impute_transcript()</a></code> function can be then used to impute
the missing transcript counts using the gene counts. The
<code><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim()</a></code> function can handle the
<code>SingleCellExperiment</code> object with the combined gene and
transcript counts, with or without imputed values, <code>NA</code>
values will be plotted in grey.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_sce.html">combine_sce</a></span><span class="op">(</span><span class="va">sce_lr</span>, <span class="va">scmixology_lib90</span><span class="op">)</span></span>
<span><span class="va">combined_sce</span> <span class="op">&lt;-</span> <span class="va">combined_sce</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># plot without imputation</span></span>
<span><span class="fu"><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim</a></span><span class="op">(</span><span class="va">combined_sce</span>, <span class="st">'ENSG00000108107'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMES_vignette_files/figure-html/unnamed-chunk-5-1.png" width="540"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># impute missing transcript counts</span></span>
<span><span class="va">combined_sce_impute</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sc_impute_transcript.html">sc_impute_transcript</a></span><span class="op">(</span><span class="va">combined_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; Imputing transcript counts ...</span></span>
<span><span class="fu"><a href="../reference/plot_isoform_reduced_dim.html">plot_isoform_reduced_dim</a></span><span class="op">(</span><span class="va">combined_sce_impute</span>, <span class="st">'ENSG00000108107'</span><span class="op">)</span></span></code></pre></div>
<p><img src="FLAMES_vignette_files/figure-html/unnamed-chunk-5-2.png" width="540"></p>
</div>
<div class="section level4">
<h4 id="flames-termination">FLAMES termination<a class="anchor" aria-label="anchor" href="#flames-termination"></a>
</h4>
<p>The directory <code>outdir</code> now contains several output files
returned from this pipeline. The output files generated by this pipeline
are:</p>
<ul>
<li>
<code>transcript_count.csv.gz</code> - a transcript count matrix
(also contained in the output <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a>
or <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>)</li>
<li>
<code>isoform_annotated.filtered.gff3</code> - found isoform
information in gff3 format</li>
<li>
<code>transcript_assembly.fa</code> - transcript sequence from the
isoforms</li>
<li>
<code>align2genome.bam</code> sorted BAM file with reads aligned to
genome (intermediate FLAMES step)</li>
<li>
<code>realign2transcript.bam</code> - sorted realigned BAM file
using the transcript_assembly.fa as reference (intermediate FLAMES
step)</li>
<li>
<code>tss_tes.bedgraph</code>- TSS TES enrichment for all reads (for
QC)</li>
</ul>
<p>The pipeline also returns a <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a>
or <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object, depending on the pipeline run, containing the data from
<code>transcript_count.csv.gz</code>and
<code>isoform_annotated.filtered.gff3</code> <span class="citation">(Amezquita et al. 2020)</span> <span class="citation">(Morgan et al. 2020)</span>. This <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a>
(or <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>)
object contains the same data as present in the <code>outdir</code>
directory, and is given to simplify the process of reading the
transcript count matrix and annotation data back into R, for further
analysis.</p>
</div>
</div>
<div class="section level3">
<h3 id="flames-bulk-pipeline">FLAMES Bulk Pipeline<a class="anchor" aria-label="anchor" href="#flames-bulk-pipeline"></a>
</h3>
<p>A basic example of the execution of the FLAMES bulk pipeline is given
below. The process for this is essentially identical to the above
example for single cell data.</p>
<div class="section level4">
<h4 id="environment-setup-1">Environment setup<a class="anchor" aria-label="anchor" href="#environment-setup-1"></a>
</h4>
<p>To get started, the pipeline needs access to a gene annotation file
in GFF3 or GTF format, a directory containing one or more FASTQ files
(which will be merged as pre-processing), a genome FASTA file, as well
as the file path to minimap2, and the file path to the directory to hold
output files.</p>
<p>For this example, these files are downloaded from GitHub using <a href="http://bioconductor.org/packages/release/bioc/html/BiocFileCache.html" class="external-link">BiocFileCache</a>
<span class="citation">(Shepherd and Morgan 2021)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">temp_path</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">file_url</span> <span class="op">&lt;-</span></span>
<span>  <span class="st">"https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data"</span></span>
<span></span>
<span><span class="va">annot</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>  <span class="va">bfc</span>, <span class="st">"Annotation"</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">genome_fa</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>  <span class="va">bfc</span>, <span class="st">"Genomefa"</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"SIRV_isoforms_multi-fasta_170612a.fasta"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># download the two fastq files, move them to a folder to be merged together</span></span>
<span><span class="va">fastq1</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">"Fastq1"</span>, </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"fastq"</span>, <span class="st">"sample1.fastq.gz"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fastq2</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">"Fastq2"</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">file_url</span>, <span class="st">"fastq"</span>, <span class="st">"sample2.fastq.gz"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># the downloaded fastq files need to be in a directory to be merged together</span></span>
<span><span class="va">fastq_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">temp_path</span>, <span class="st">"fastq_dir"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">fastq_dir</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">fastq1</span>, <span class="va">fastq2</span><span class="op">)</span>, <span class="va">fastq_dir</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">fastq1</span>, <span class="va">fastq2</span><span class="op">)</span><span class="op">)</span> <span class="co"># the original files can be deleted</span></span>
<span></span>
<span><span class="co"># setup other environment variables</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span><span class="va">config_file</span> <span class="op">&lt;-</span> <span class="fu">FLAMES</span><span class="fu">::</span><span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span><span class="co">#&gt; Writing configuration parameters to:  /tmp/RtmpsYpfbT/file1d78358a0e9b/config_file_7544.json</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="flames-execution-1">FLAMES execution<a class="anchor" aria-label="anchor" href="#flames-execution-1"></a>
</h4>
<p>Once the environment has been setup, the pipeline can be executed by
running:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mritchielab.github.io/FLAMES">FLAMES</a></span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="../reference/sys_which.html">sys_which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"minimap2"</span>, <span class="st">"k8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">summarizedExperiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bulk_long_pipeline.html">bulk_long_pipeline</a></span><span class="op">(</span></span>
<span>    annot <span class="op">=</span> <span class="va">annot</span>, fastq <span class="op">=</span> <span class="va">fastq_dir</span>, outdir <span class="op">=</span> <span class="va">outdir</span>,</span>
<span>    genome_fa <span class="op">=</span> <span class="va">genome_fa</span>, config_file <span class="op">=</span> <span class="va">config_file</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="flames-termination-1">FLAMES termination<a class="anchor" aria-label="anchor" href="#flames-termination-1"></a>
</h4>
<p>After the bulk pipeline has completed, the output directory contains
the same files as the single cell pipeline produces.
<code>bulk_long_pipeline</code> also returns a <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a>
object, containing the same data as the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
as above <span class="citation">(Amezquita et al. 2020)</span> <span class="citation">(Morgan et al. 2020)</span>.</p>
</div>
</div>
<div class="section level3">
<h3 id="running-with-slurm">Running with Slurm<a class="anchor" aria-label="anchor" href="#running-with-slurm"></a>
</h3>
<p>Since the barcode demultiplexing step, isoform identification step
and quantification step is currently single threaded, job scheduler
(such as Slurm) users may want to allocate different resources for each
step. This can be achieved by calling the individual functions
(<code>find_isoform</code>, <code>minimap2_realign</code> etc.)
sequentially, or by changing the configuration file. The
<code>Bash</code> script below demonstrates how this can be done with a
single <code>R</code> script calling the pipeline function.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co">#!/bin/bash -x</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># set all steps to false</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/"do_/s/true/false/'</span> config_sclr_nanopore_3end.json </span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># set do_genome_alignment to true</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/do_genome_alignment/s/false/true/'</span> <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>        config_sclr_nanopore_3end.json </span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="ex">srun</span> <span class="at">-c</span> 20 <span class="at">--mem</span> 64GB <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>        Rscript flames_pipeline.R <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/"do_/s/true/false/'</span> <span class="dt">\</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/do_isoform_identification/s/false/true/'</span> <span class="dt">\</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="ex">srun</span> <span class="at">-c</span> 1 <span class="at">--mem</span> 64GB <span class="at">--ntasks</span><span class="op">=</span>1 <span class="dt">\</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>        Rscript flames_pipeline.R  <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/"do_/s/true/false/'</span> <span class="dt">\</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/do_read_realignment/s/false/true/'</span> <span class="dt">\</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="ex">srun</span> <span class="at">-c</span> 20 <span class="at">--mem</span> 64GB <span class="at">--ntasks</span><span class="op">=</span>1 <span class="dt">\</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>        Rscript flames_pipeline.R <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/"do_/s/true/false/'</span> <span class="dt">\</span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">'/do_transcript_quantification/s/false/true/'</span> <span class="dt">\</span></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>        config_sclr_nanopore_3end.json <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="ex">srun</span> <span class="at">-c</span> 1 <span class="at">--mem</span> 64GB <span class="at">--ntasks</span><span class="op">=</span>1 <span class="dt">\</span></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a>        Rscript flames_pipeline.R <span class="kw">&amp;&amp;</span> <span class="dt">\</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Pipeline finished"</span></span></code></pre></div>
<p>The <code>Bash</code> script can then be executed inside a tmux or
screen session with:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">./flames.sh</span> <span class="kw">|</span> <span class="fu">tee</span> <span class="at">-a</span> bash.log<span class="kw">;</span> <span class="fu">tail</span> bash.log <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>        <span class="kw">|</span> <span class="ex">mail</span> <span class="at">-s</span> <span class="st">"flames pipeline ended"</span> user@example.com</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="flames-on-windows">FLAMES on Windows<a class="anchor" aria-label="anchor" href="#flames-on-windows"></a>
</h3>
<p>Due to FLAMES requiring minimap2 and pysam, FLAMES is currently
unavaliable on Windows.</p>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>Please cite flames’s paper if you use flames in your research. As
FLAMES used incorporates BLAZE, flexiplex and minimap2, samtools, bambu.
Please make sure to cite when using these tools.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [3] Biobase_2.64.0              GenomicRanges_1.56.1       </span></span>
<span><span class="co">#&gt;  [5] GenomeInfoDb_1.40.1         IRanges_2.38.1             </span></span>
<span><span class="co">#&gt;  [7] S4Vectors_0.42.1            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt;  [9] MatrixGenerics_1.16.0       matrixStats_1.4.1          </span></span>
<span><span class="co">#&gt; [11] FLAMES_1.11.1               BiocStyle_2.32.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.4                  ProtGenerics_1.36.0      </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-8              httr_1.4.7               </span></span>
<span><span class="co">#&gt;   [5] RColorBrewer_1.1-3        doParallel_1.0.17        </span></span>
<span><span class="co">#&gt;   [7] tools_4.4.1               backports_1.5.0          </span></span>
<span><span class="co">#&gt;   [9] utf8_1.2.4                R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [11] HDF5Array_1.32.1          lazyeval_0.2.2           </span></span>
<span><span class="co">#&gt;  [13] uwot_0.2.2                rhdf5filters_1.16.0      </span></span>
<span><span class="co">#&gt;  [15] GetoptLong_1.0.5          withr_3.0.1              </span></span>
<span><span class="co">#&gt;  [17] prettyunits_1.2.0         GGally_2.2.1             </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3             cli_3.6.3                </span></span>
<span><span class="co">#&gt;  [21] textshaping_0.4.0         labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [23] ggbio_1.52.0              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [25] pkgdown_2.1.0.9000        Rsamtools_2.20.0         </span></span>
<span><span class="co">#&gt;  [27] systemfonts_1.1.0         txdbmaker_1.0.1          </span></span>
<span><span class="co">#&gt;  [29] foreign_0.8-87            R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [31] dichromat_2.0-0.1         scater_1.32.1            </span></span>
<span><span class="co">#&gt;  [33] parallelly_1.38.0         BSgenome_1.72.0          </span></span>
<span><span class="co">#&gt;  [35] limma_3.60.4              rstudioapi_0.16.0        </span></span>
<span><span class="co">#&gt;  [37] RSQLite_2.3.7             FNN_1.1.4                </span></span>
<span><span class="co">#&gt;  [39] generics_0.1.3            shape_1.4.6.1            </span></span>
<span><span class="co">#&gt;  [41] BiocIO_1.14.0             dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [43] Matrix_1.7-0              ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;  [45] fansi_1.0.6               abind_1.4-8              </span></span>
<span><span class="co">#&gt;  [47] R.methodsS3_1.8.2         lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.10               edgeR_4.2.1              </span></span>
<span><span class="co">#&gt;  [51] rhdf5_2.48.0              SparseArray_1.4.8        </span></span>
<span><span class="co">#&gt;  [53] BiocFileCache_2.12.0      grid_4.4.1               </span></span>
<span><span class="co">#&gt;  [55] blob_1.2.4                dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [57] crayon_1.5.3              dir.expiry_1.12.0        </span></span>
<span><span class="co">#&gt;  [59] lattice_0.22-6            beachmat_2.20.0          </span></span>
<span><span class="co">#&gt;  [61] cowplot_1.1.3             GenomicFeatures_1.56.0   </span></span>
<span><span class="co">#&gt;  [63] KEGGREST_1.44.1           pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [65] knitr_1.48                ComplexHeatmap_2.20.0    </span></span>
<span><span class="co">#&gt;  [67] metapod_1.12.0            bambu_3.6.0              </span></span>
<span><span class="co">#&gt;  [69] rjson_0.2.22              xgboost_1.7.8.1          </span></span>
<span><span class="co">#&gt;  [71] codetools_0.2-20          glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [73] data.table_1.16.0         vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [75] png_0.1-8                 gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [77] cachem_1.1.0              xfun_0.47                </span></span>
<span><span class="co">#&gt;  [79] S4Arrays_1.4.1            DropletUtils_1.24.0      </span></span>
<span><span class="co">#&gt;  [81] iterators_1.0.14          statmod_1.5.0            </span></span>
<span><span class="co">#&gt;  [83] bluster_1.14.0            bit64_4.0.5              </span></span>
<span><span class="co">#&gt;  [85] progress_1.2.3            filelock_1.0.3           </span></span>
<span><span class="co">#&gt;  [87] bslib_0.8.0               irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [89] vipor_0.4.7               rpart_4.1.23             </span></span>
<span><span class="co">#&gt;  [91] colorspace_2.1-1          DBI_1.2.3                </span></span>
<span><span class="co">#&gt;  [93] Hmisc_5.1-3               nnet_7.3-19              </span></span>
<span><span class="co">#&gt;  [95] tidyselect_1.2.1          bit_4.0.5                </span></span>
<span><span class="co">#&gt;  [97] compiler_4.4.1            curl_5.2.2               </span></span>
<span><span class="co">#&gt;  [99] httr2_1.0.4               graph_1.82.0             </span></span>
<span><span class="co">#&gt; [101] htmlTable_2.4.3           BiocNeighbors_1.22.0     </span></span>
<span><span class="co">#&gt; [103] basilisk.utils_1.16.0     xml2_1.3.6               </span></span>
<span><span class="co">#&gt; [105] desc_1.4.3                DelayedArray_0.30.1      </span></span>
<span><span class="co">#&gt; [107] bookdown_0.40             rtracklayer_1.64.0       </span></span>
<span><span class="co">#&gt; [109] checkmate_2.3.2           scales_1.3.0             </span></span>
<span><span class="co">#&gt; [111] RBGL_1.80.0               rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt; [113] stringr_1.5.1             digest_0.6.37            </span></span>
<span><span class="co">#&gt; [115] rmarkdown_2.28            basilisk_1.16.0          </span></span>
<span><span class="co">#&gt; [117] XVector_0.44.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt; [119] pkgconfig_2.0.3           base64enc_0.1-3          </span></span>
<span><span class="co">#&gt; [121] sparseMatrixStats_1.16.0  highr_0.11               </span></span>
<span><span class="co">#&gt; [123] dbplyr_2.5.0              fastmap_1.2.0            </span></span>
<span><span class="co">#&gt; [125] ensembldb_2.28.1          rlang_1.1.4              </span></span>
<span><span class="co">#&gt; [127] GlobalOptions_0.1.2       htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt; [129] UCSC.utils_1.0.0          DelayedMatrixStats_1.26.0</span></span>
<span><span class="co">#&gt; [131] farver_2.1.2              jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt; [133] jsonlite_1.8.8            BiocParallel_1.38.0      </span></span>
<span><span class="co">#&gt; [135] R.oo_1.26.0               BiocSingular_1.20.0      </span></span>
<span><span class="co">#&gt; [137] VariantAnnotation_1.50.0  RCurl_1.98-1.16          </span></span>
<span><span class="co">#&gt; [139] magrittr_2.0.3            Formula_1.2-5            </span></span>
<span><span class="co">#&gt; [141] scuttle_1.14.0            GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt; [143] Rhdf5lib_1.26.0           munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [145] Rcpp_1.0.13               viridis_0.6.5            </span></span>
<span><span class="co">#&gt; [147] reticulate_1.39.0         stringi_1.8.4            </span></span>
<span><span class="co">#&gt; [149] zlibbioc_1.50.0           plyr_1.8.9               </span></span>
<span><span class="co">#&gt; [151] ggstats_0.6.0             parallel_4.4.1           </span></span>
<span><span class="co">#&gt; [153] listenv_0.9.1             ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt; [155] Biostrings_2.72.1         hms_1.1.3                </span></span>
<span><span class="co">#&gt; [157] circlize_0.4.16           locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt; [159] igraph_2.0.3              reshape2_1.4.4           </span></span>
<span><span class="co">#&gt; [161] biomaRt_2.60.1            ScaledMatrix_1.12.0      </span></span>
<span><span class="co">#&gt; [163] XML_3.99-0.17             evaluate_0.24.0          </span></span>
<span><span class="co">#&gt; [165] biovizBase_1.52.0         scran_1.32.0             </span></span>
<span><span class="co">#&gt; [167] BiocManager_1.30.25       foreach_1.5.2            </span></span>
<span><span class="co">#&gt; [169] tidyr_1.3.1               purrr_1.0.2              </span></span>
<span><span class="co">#&gt; [171] future_1.34.0             clue_0.3-65              </span></span>
<span><span class="co">#&gt; [173] ggplot2_3.5.1             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt; [175] restfulr_0.0.15           AnnotationFilter_1.28.0  </span></span>
<span><span class="co">#&gt; [177] viridisLite_0.4.2         ragg_1.3.3               </span></span>
<span><span class="co">#&gt; [179] OrganismDbi_1.46.0        tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [181] memoise_2.0.1             beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt; [183] AnnotationDbi_1.66.0      GenomicAlignments_1.40.0 </span></span>
<span><span class="co">#&gt; [185] cluster_2.1.6             globals_0.16.3</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-singlecellexperiment" class="csl-entry">
Amezquita, Robert, Aaron Lun, Etienne Becht, Vince Carey, Lindsay Carpp,
Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“Orchestrating
Single-Cell Analysis with Bioconductor.”</span> <em>Nature Methods</em>
17: 137–45. <a href="https://www.nature.com/articles/s41592-019-0654-x" class="external-link">https://www.nature.com/articles/s41592-019-0654-x</a>.
</div>
<div id="ref-bambu" class="csl-entry">
Chen, Ying, Andre Sim, Yuk Kei Wan, Keith Yeo, Joseph Jing Xian Lee, Min
Hao Ling, Michael I Love, and Jonathan Göke. 2023. <span>“Context-Aware
Transcript Quantification from Long-Read RNA-Seq Data with
Bambu.”</span> <em>Nature Methods</em>, 1–9.
</div>
<div id="ref-flexiplex" class="csl-entry">
Davidson, Nadia M, Noorul Amin, Ling Min Hao, Changqing Wang, Oliver
Cheng, Jonathan M Goeke, Matthew E Ritchie, and Shuyi Wu. 2023.
<span>“Flexiplex: A Versatile Demultiplexer and Search Tool for Omics
Data.”</span> <em>bioRxiv</em>, 2023–08.
</div>
<div id="ref-minimap2" class="csl-entry">
Li, Heng. 2018. <span>“<span class="nocase">Minimap2: pairwise alignment
for nucleotide sequences</span>.”</span> <em>Bioinformatics</em> 34
(18): 3094–3100. <a href="https://doi.org/10.1093/bioinformatics/bty191" class="external-link">https://doi.org/10.1093/bioinformatics/bty191</a>.
</div>
<div id="ref-summarizedexperiment" class="csl-entry">
Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020.
<em>SummarizedExperiment: SummarizedExperiment Container</em>. <a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">https://bioconductor.org/packages/SummarizedExperiment</a>.
</div>
<div id="ref-biocfilecache" class="csl-entry">
Shepherd, Lori, and Martin Morgan. 2021. <em>BiocFileCache: Manage Files
Across Sessions</em>.
</div>
<div id="ref-flames" class="csl-entry">
Tian, Luyi, Jafar S. Jabbari, Rachel Thijssen, Quentin Gouil, Shanika L.
Amarasinghe, Hasaru Kariyawasam, Shian Su, et al. 2020.
<span>“Comprehensive Characterization of Single Cell Full-Length
Isoforms in Human and Mouse with Long-Read Sequencing.”</span>
<em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.08.10.243543" class="external-link">https://doi.org/10.1101/2020.08.10.243543</a>.
</div>
<div id="ref-blaze" class="csl-entry">
You, Yupei, Yair DJ Prawer, De Paoli-Iseppi, Cameron PJ Hunt, Clare L
Parish, Heejung Shim, Michael B Clark, et al. 2023.
<span>“Identification of Cell Barcodes from Long-Read Single-Cell
RNA-Seq with BLAZE.”</span> <em>Genome Biology</em> 24 (1): 1–23.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luyi Tian, Changqing Wang, Yupei You, Oliver Voogd, Jakob Schuster, Shian Su.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
