# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

find_isoform_multithread <- function(gff3, genome_bam, isoform_gff3, tss_tes_stat, genomefa, transcript_fa, isoform_parameters, raw_splice_isoform) {
    invisible(.Call(`_FLAMES_find_isoform_multithread`, gff3, genome_bam, isoform_gff3, tss_tes_stat, genomefa, transcript_fa, isoform_parameters, raw_splice_isoform))
}

#' Rcpp port of flexiplex
#'
#' @description demultiplex reads with flexiplex, for detailed description, see
#' documentation for the original flexiplex: https://davidsongroup.github.io/flexiplex
#'
#' @param reads_in Input FASTQ or FASTA file
#' @param barcodes_file barcode allow-list file
#' @param bc_as_readid bool, whether to add the demultiplexed barcode to the
#' read ID field 
#' @param max_bc_editdistance max edit distance for barcode '
#' @param max_flank_editdistance max edit distance for the flanking sequences '
#' @param pattern StringVector defining the barcode structure, see [find_barcode]
#' @param reads_out output file for demultiplexed reads
#' @param stats_out output file for demultiplexed stats
#' @param n_threads number of threads to be used during demultiplexing
#' @param bc_out WIP
#' @return integer return value. 0 represents normal return.
#' @export
flexiplex <- function(reads_in, barcodes_file, bc_as_readid, max_bc_editdistance, max_flank_editdistance, pattern, reads_out, stats_out, bc_out, n_threads) {
    .Call(`_FLAMES_flexiplex`, reads_in, barcodes_file, bc_as_readid, max_bc_editdistance, max_flank_editdistance, pattern, reads_out, stats_out, bc_out, n_threads)
}

variant_count_matrix <- function(bam_path, seqname, pos, indel, barcodes) {
    .Call(`_FLAMES_variant_count_matrix`, bam_path, seqname, pos, indel, barcodes)
}

