#' Pipeline for Single Cell Data
#'
#' @md
#'
#' @description
#' Semi-supervised isoform detection and annotation for long read data.
#' This varient is for single cell data. By default, this pipeline demultiplexes input
#' fastq data (\code{match_cell_barcode = TRUE}). Specific parameters relating to
#' analysis can be changed either through function arguments, or through a
#' configuration JSON file.
#'
#' @details
#' By default FLAMES use minimap2 for read alignment. After the genome alignment step (\code{do_genome_align}), FLAMES summarizes the alignment for each read by grouping reads
#' with similar splice junctions to get a raw isoform annotation (\code{do_isoform_id}). The raw isoform
#' annotation is compared against the reference annotation to correct potential splice site
#' and transcript start/end errors. Transcripts that have similar splice junctions
#' and transcript start/end to the reference transcript are merged with the
#' reference. This process will also collapse isoforms that are likely to be truncated
#' transcripts. Next is the read realignment step (\code{do_read_realign}), where the sequence of each polished transcript is extracted and used as
#' the updated reference. The reads are realigned to this reference by minimap2. The
#' transcripts with only a few full-length aligned reads are discarded.
#' The reads are assigned to transcripts based on both alignment score, fractions of
#' reads aligned and transcript coverage. Reads that cannot be uniquely assigned to
#' transcripts or have low transcript coverage are discarded. The UMI transcript
#' count matrix is generated by collapsing the reads with the same UMI in a similar
#' way to what is done for short-read scRNA-seq data, but allowing for an edit distance
#' of up to 2 by default. Most of the parameters, such as the minimal distance to splice site and minimal percentage of transcript coverage
#' can be modified by the JSON configuration file (\code{config_file}).
#'
#' @param annot gene annotations file in gff3  format
#' @param fastq file path to input fastq file
#' @param in_bam optional bam file to use instead of fastq file
#' @param outdir directory to store all output files.
#' @param genome_fa genome fasta file.
#' @param minimap2_dir directory containing minimap2, k8 and paftools.js program.
#' k8 and paftools.js are used to convert gff3 to bed12.
#' @param downsample_ratio downsampling ratio if performing downsampling analysis.
#' @param config_file JSON configuration file. If specified, \code{config_file} overrides
#' all configuration parameters
#' @param do_genome_align Boolean. Specifies whether to run the genome alignment step. \code{TRUE} is recommended
#' @param do_isoform_id Boolean. Specifies whether to run the isoform identification step. \code{TRUE} is recommended
#' @param do_read_realign Boolean. Specifies whether to run the read realignment step. \code{TRUE} is recommended
#' @param do_transcript_quanti Boolean. Specifies whether to run the transcript quantification step. \code{TRUE} is recommended
#' @param gen_raw_isoform Boolean.
#' @param has_UMI Boolean. Specifies if the data contains UMI.
#' @param MAX_DIST Maximum distance allowed when merging splicing sites in isoform consensus clustering.
#' @param MAX_TS_DIST Maximum distance allowed when merging transcript start/end position in isoform consensus clustering.
#' @param MAX_SPLICE_MATCH_DIST Maximum distance allowed when merging splice site called from the data and the reference annotation.
#' @param min_fl_exon_len Minimum length for the first exon outside the gene body in reference annotation. This is to correct the alignment artifact
#' @param Max_site_per_splice Maximum transcript start/end site combinations allowed per splice chain
#' @param Min_sup_cnt Minimum number of read support an isoform. Decreasing this number will significantly increase the number of isoform detected.
#' @param Min_cnt_pct Minimum percentage of count for an isoform relative to total count for the same gene.
#' @param Min_sup_pct Minimum percentage of count for an splice chain that support a given transcript start/end site combination.
#' @param strand_specific 1, -1 or 0. 1 indicates if reads are in the same
#' strand as mRNA, -1 indicates reads are reverse complemented, 0 indicates
#' reads are not strand specific.
#' @param remove_incomp_reads The strenge of truncated isoform filtering. larger number means more stringent filtering.
#' @param use_junctions whether to use known splice junctions to help correct the alignment results
#' @param no_flank Boolean. for synthetic spike-in data. refer to Minimap2 document for detail
#' @param use_annotation Boolean. whether to use reference to help annotate known isoforms
#' @param min_tr_coverage Minimum percentage of isoform coverage for a read to be aligned to that isoform
#' @param min_read_coverage Minimum percentage of read coverage for a read to be uniquely aligned to that isoform
#' @param UMI_LEN The length of UMI sequence
#' @param reference_csv file path to the reference csv used for demultiplexing
#' @param match_barcode bool; specifies if demultiplexing should be performed using `FLAMES::match_cell_barcode`
#' @return \code{sc_long_pipeline} returns a SingleCellExperiment object, containing a count
#' matrix as an assay, gene annotations under metadata, as well as a list of the other
#' output files generated by the pipeline. The pipeline also outputs a number of output
#' files into the given \code{outdir} directory. These output files generated by the pipeline are:
#' \itemize{
#'  \item{transcript_count.csv.gz}{ - a transcript count matrix (also contained in the SingleCellExperiment)}
#'  \item{isoform_annotated.filtered.gff3}{ - isoforms in gff3 format (also contained in the SingleCellExperiment)}
#'  \item{transcript_assembly.fa}{ - transcript sequence from the isoforms}
#'  \item{align2genome.bam}{ - sorted BAM file with reads aligned to genome}
#'  \item{realign2transcript.bam}{ - sorted realigned BAM file using the transcript_assembly.fa as reference}
#'  \item{tss_tes.bedgraph}{ - TSS TES enrichment for all reads (for QC)}
#' }
#'
#' @details The default parameters can be changed either through the function
#' arguments are through the configuration JSON file \code{config_file}. the \code{pipeline_parameters}
#' section specifies which steps are to be executed in the pipeline - by default, all
#' steps are executed. The \code{isoform_parameters} section affects isoform detection - key
#' parameters include:
#' \itemize{
#'  \item{\code{Min_sup_cnt}}{ which causes transcripts with less reads aligned than
#' it's value to be discarded}
#'  \item{\code{MAX_TS_DIST}}{ which merges transcripts with the same intron
#' chain and TSS/TES distace less than \code{MAX_TS_DIST}}
#'  \item{\code{strand_specific}}{ which specifies if reads are in the same strand as the mRNA (1),
#' or the reverse complemented (-1) or not strand specific (0), which results in
#' strand information being based on reference annotation.}
#' }
#'
#' @seealso
#' [bulk_long_pipeline()] for bulk long data,
#' [SingleCellExperiment()] for how data is outputted
#'
#' @importFrom SingleCellExperiment SingleCellExperiment
#' @importFrom utils read.csv read.table
#'
#' @example inst/examples/pipeline_example.R
#' @export
sc_long_pipeline <-
    function(annot,
             fastq,
             in_bam = NULL,
             outdir,
             genome_fa,
             minimap2_dir = NULL,
             downsample_ratio = 1,
             reference_csv,
             match_barcode = TRUE,
             config_file = NULL,
             do_genome_align = TRUE,
             do_isoform_id = TRUE,
             do_read_realign = TRUE,
             do_transcript_quanti = TRUE,
             gen_raw_isoform = TRUE,
             has_UMI = FALSE,
             UMI_LEN = 10,
             MAX_DIST = 10,
             MAX_TS_DIST = 100,
             MAX_SPLICE_MATCH_DIST = 10,
             min_fl_exon_len = 40,
             Max_site_per_splice = 3,
             Min_sup_cnt = 10,
             Min_cnt_pct = 0.01,
             Min_sup_pct = 0.2,
             strand_specific = 1,
             remove_incomp_reads = 5,
             use_junctions = TRUE,
             no_flank = TRUE,
             use_annotation = TRUE,
             min_tr_coverage = 0.75,
             min_read_coverage = 0.75) {
        if (is.null(in_bam)) {
            if (match_barcode) {
                if (!file.exists(reference_csv)) {
                      stop("reference_csv must exists.")
                  }
                infq <-
                    paste(outdir, "matched_reads.fastq.gz", sep = "/")
                bc_stat <-
                    paste(outdir, "matched_barcode_stat", sep = "/")
                match_cell_barcode(
                    fastq,
                    bc_stat,
                    infq,
                    reference_csv,
                    MAX_DIST,
                    UMI_LEN
                )
            } else {
                  infq <- fastq
              } # requesting to not match barcodes implies `fastq` has already been run through the
            # function in a previous FLAMES call
        }

        out_files <-
            generic_long_pipeline(
                annot,
                infq,
                in_bam = in_bam,
                outdir,
                genome_fa,
                minimap2_dir,
                downsample_ratio,
                config_file,
                do_genome_align,
                do_isoform_id,
                do_read_realign,
                do_transcript_quanti,
                gen_raw_isoform,
                has_UMI,
                MAX_DIST,
                MAX_TS_DIST,
                MAX_SPLICE_MATCH_DIST,
                min_fl_exon_len,
                Max_site_per_splice,
                Min_sup_cnt,
                Min_cnt_pct,
                Min_sup_pct,
                strand_specific,
                remove_incomp_reads,
                use_junctions,
                no_flank,
                use_annotation,
                min_tr_coverage,
                min_read_coverage
            )

        sce <- generate_sc_singlecell(out_files)

        sce
    }

generate_sc_singlecell <- function(out_files) {
    # this method requires testing using single cell data
    counts <- read.csv(out_files$counts)
    annot <- read.table(out_files$annot)
    colnames(annot) <-
        c(
            "SequenceID",
            "Source",
            "Feature",
            "Start",
            "End",
            "Score",
            "Strand",
            "Phase",
            "Attributes"
        )
    mdata <- list(
        "Annotations" = out_files$annot,
        "OutputFiles" = out_files
    )
    sce <-
        SingleCellExperiment::SingleCellExperiment(list(
            "Flames Single Cell" =
                counts
        ),
        metadata = mdata
        )
    # return the created singlecellexperiment
    sce
}
